### PR TITLE
Add lockfile path

### PR DESCRIPTION
**Pull request by ejortega**
Add updates to submit the lockfile path associated with each package.

Closes https://github.com/phylum-dev/cli/issues/1152

## Checklist

- [x] Does this PR have an associated issue (i.e., `closes #<issueNum>` in description above)?
- [x] Have you ensured that you have met the expected acceptance criteria?
- [ ] Have you created sufficient tests?
- [ ] Have you updated all affected documentation?
- [x] Have you updated CHANGELOG.md (or extensions/CHANGELOG.md), if applicable

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -38,7 +38,8 @@ maplit = "1.0.2"
 open = "5.0.0"
 phylum_lockfile = { path = "../lockfile", features = ["generator"] }
 phylum_project = { path = "../phylum_project" }
-phylum_types = { git = "https://github.com/phylum-dev/phylum-types", branch = "development" }
+phylum_types = { git = "https://github.com/phylum-dev/phylum-types", branch = "add-lockfile-path" }
```
Comment (ejortega) [inline cli/Cargo.toml:41]: Update back to development when PR is merged.

File: lockfile/Cargo.toml [hunk 2]
```diff
@@ -16,8 +16,8 @@ ignore = "0.4.20"
 lockfile_generator = { path = "../lockfile_generator", optional = true }
 log = "0.4.6"
 nom = "7.1.1"
-packageurl = "0.3.0"
-phylum_types = { git = "https://github.com/phylum-dev/phylum-types", branch = "development" }
+purl = "0.1.1"
+phylum_types = { git = "https://github.com/phylum-dev/phylum-types", branch = "add-lockfile-path" }
```
Comment (ejortega) [inline lockfile/Cargo.toml:20]: Update back to development when PR is merged.

File: phylum_project/Cargo.toml [hunk 3]
```diff
@@ -9,7 +9,7 @@ edition = "2021"
 rust-version = "1.64.0"
 
 [dependencies]
-phylum_types = { git = "https://github.com/phylum-dev/phylum-types", branch = "development" }
+phylum_types = { git = "https://github.com/phylum-dev/phylum-types", branch = "add-lockfile-path" }
```
Comment (ejortega) [inline phylum_project/Cargo.toml:12]: Update back to development when PR is merged.

File: cli/src/commands/jobs.rs [hunk 4]
```diff
@@ -158,12 +161,13 @@ pub async fn handle_submission(api: &mut PhylumApi, matches: &clap::ArgMatches)
...
+                        .with_version(pkg_version)
+                        .build()?
+                        .to_string();
 
-                    packages.push(PackageDescriptor {
-                        name: pkg_name.to_owned(),
-                        version: pkg_version.to_owned(),
-                        package_type: request_type.to_owned(),
-                    });
+                    packages.push(ParsedPurl { purl, lockfile_path: None });
```
Comment (ejortega) [inline cli/src/commands/jobs.rs:170]: Guess the path to file used for batch submissions can be included as a lock file path?

File: cli/src/api/mod.rs [hunk 5]
```diff
@@ -566,9 +583,12 @@ mod tests {
...
                         {
-                            "name": "ansi-red",
-                            "type": "npm",
-                            "version": "0.1.1"
+                            "package_descriptor": {
+                                "name": "ansi-red",
+                                "version": "0.1.1",
+                                "type": "npm"
+                            },
+                            "lockfile_path": null
```
Comment (kylewillmon) [inline cli/src/api/mod.rs:591]: This has been [flattened](https://github.com/phylum-dev/phylum-types/pull/78/commits/7c50567398d9303e700fcef734d8b6b250daf501) now, right?

If we also use `#[skip_serializing_if = "Option::is_none"]`, we can avoid any changes to the format when no `lockfile_path` is used.

File: cli/src/commands/jobs.rs [hunk 6]
```diff
@@ -158,11 +162,15 @@ pub async fn handle_submission(api: &mut PhylumApi, matches: &clap::ArgMatches)
...
+                    let pkg_descriptor = PackageDescriptor {
                         name: pkg_name.to_owned(),
                         version: pkg_version.to_owned(),
                         package_type: request_type.to_owned(),
+                    };
+
+                    packages.push(PackageDescriptorAndLockfilePath {
+                        package_descriptor: pkg_descriptor,
+                        lockfile_path: None,
                     });
```
Comment (kylewillmon) [inline cli/src/commands/jobs.rs:174]: You could make your life easier with a `From` impl.

```rust
impl From<PackageDescriptor> for PackageDescriptorAndLockfilePath {
    fn from(package_descriptor: PackageDescriptor) -> Self {
        Self {
            package_descriptor,
            lockfile_path: None,
        }
    }
}
```

And then this code becomes:

```suggestion
                    packages.push(PackageDescriptor {
                        name: pkg_name.to_owned(),
                        version: pkg_version.to_owned(),
                        package_type: request_type.to_owned(),
                    }.into());
```

File: cli/src/commands/parse.rs [hunk 7]
```diff
@@ -17,6 +17,19 @@ pub struct ParsedLockfile {
     pub packages: Vec<PackageDescriptor>,
 }
 
+impl From<ParsedLockfile> for Vec<PackageDescriptorAndLockfilePath> {
+    fn from(value: ParsedLockfile) -> Self {
+        let mut packages = Vec::with_capacity(value.packages.len());
+        for pkg_descriptor in value.packages {
+            packages.push(PackageDescriptorAndLockfilePath {
+                package_descriptor: pkg_descriptor,
+                lockfile_path: Some(value.path.to_string_lossy().to_string()),
```
Comment (kylewillmon) [inline cli/src/commands/parse.rs:26]: `Cow::into_owned` is more efficient here (because `.to_string()` always copies while `.into_owned()` only copies if needed).

```suggestion
                lockfile_path: Some(value.path.to_string_lossy().into_owned()),
```

File: cli/src/commands/parse.rs [hunk 8]
```diff
@@ -31,10 +44,12 @@ pub fn lockfile_types(add_auto: bool) -> Vec<&'static str> {
...
     let lockfiles = config::lockfiles(matches, phylum_project::get_current_project().as_ref())?;
 
-    let mut pkgs: Vec<PackageDescriptor> = Vec::new();
+    let mut pkgs: Vec<PackageDescriptorAndLockfilePath> = Vec::new();
 
     for lockfile in lockfiles {
-        pkgs.extend(parse_lockfile(lockfile.path, Some(&lockfile.lockfile_type))?.packages);
+        let parsed_lockfile = parse_lockfile(lockfile.path, Some(&lockfile.lockfile_type))?;
+        let parsed_packages = Vec::<PackageDescriptorAndLockfilePath>::from(parsed_lockfile);
+        pkgs.extend(parsed_packages);
```
Comment (kylewillmon) [inline cli/src/commands/parse.rs:52]: This seems unnecessary... Rust should be able to figure this out...

```suggestion
        pkgs.extend(parsed_lockfile.into());
```

If for some reason it can't, it's still cleaner to just specify the type of the variable and then call `.into()`

```suggestion
        let parse_packages: Vec<PackageDescriptorAndLockfilePath> = parse_lockfile.into();
        pkgs.extend(parse_packages);
```

File: cli/src/api/mod.rs (hunk 5 above)
Comment (ejortega) [reply cli/src/api/mod.rs:591]: It has been. Thanks, I put up a [PR](https://github.com/phylum-dev/phylum-types/pull/79) to add that attribute

File: cli/src/commands/parse.rs [hunk 9]
```diff
@@ -17,6 +17,19 @@ pub struct ParsedLockfile {
     pub packages: Vec<PackageDescriptor>,
 }
 
+impl From<ParsedLockfile> for Vec<PackageDescriptorAndLockfilePath> {
```
Comment (kylewillmon) [inline cli/src/commands/parse.rs:20]: It looks like all the places that use this immediately call `.extend()`

If that's the use case, you could instead `impl IntoIterator for ParsedLockfile` and then just call `.extend(parsed_lockfile)` without creating a `Vec` in between

File: lockfile/src/spdx.rs [hunk 10]
```diff
@@ -78,9 +78,9 @@ fn type_from_url(url: &str) -> anyhow::Result<PackageType> {
 }
 
 fn from_purl(pkg_url: &str, pkg_info: &PackageInformation) -> anyhow::Result<Package> {
-    let purl = PackageUrl::from_str(pkg_url)?;
+    let purl = GenericPurl::<String>::from_str(pkg_url)?;
```
Comment (kylewillmon) [inline lockfile/src/spdx.rs:81]: This seems like an unrelated change. Is there a reason this was done in this PR?

File: cli/src/commands/extensions/api.rs [hunk 11]
```diff
@@ -137,7 +138,7 @@ struct ProcessOutput {
 #[op]
 async fn analyze(
     op_state: Rc<RefCell<OpState>>,
-    packages: Vec<PackageDescriptor>,
+    packages: Vec<PackageDescriptorAndLockfilePath>,
```
Comment (kylewillmon) [inline cli/src/commands/extensions/api.rs:141]: This change needs:

1. An entry in `extensions/CHANGELOG.md`
2. To be documented in the docs for `PhylumApi.analyze()` in `extensions/phylum.ts`

**Review by kylewillmon — CHANGES_REQUESTED** [review]

File: lockfile/src/spdx.rs (hunk 10 above)
Comment (ejortega) [reply lockfile/src/spdx.rs:81]: Oh, I had made an initial attempt at using PURL and removed it. This got left behind, but I'll remove it.

**Review by kylewillmon — CHANGES_REQUESTED** [review]

File: extensions/phylum.ts [hunk 12]
```diff
@@ -110,14 +111,15 @@ export class PhylumApi {
...
+   *   { name: "accepts", version: "1.3.8", type: "npm", lockfile_path: "/path/to/lockfile" },
    *   { name: "ms", version: "2.0.0", type: "npm" },
    *   { name: "negotiator", version: "0.6.3", type: "npm" },
-   *   { name: "ms", version: "2.1.3", type: "npm" }
+   *   { name: "ms", version: "2.1.3", type: "npm" },
    * ]
    * ```
    *
    * Accepted package types are "npm", "pypi", "maven", "rubygems", "nuget", "cargo", and "golang"
+   * The `lockfile_path` is optional and if provided, should be a string representing the path to the lockfile.
```
[the remark this answers is not in the document]
Comment (ejortega) [reply extensions/phylum.ts:122]: It's just a differentiator of where the packages are in a repository. I don't think it has to necessarily be enforced as a path. I'm down for changing the name.

**Review by kylewillmon — APPROVED** [review]
